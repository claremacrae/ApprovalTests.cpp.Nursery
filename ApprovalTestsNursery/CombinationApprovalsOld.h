#ifndef APPROVALTESTS_CPP_COMBINATIONAPPROVALSOLD_H
#define APPROVALTESTS_CPP_COMBINATIONAPPROVALSOLD_H

#include <functional>
#include <vector>
#include <ostream>
#include "reporters/DefaultReporter.h"
#include "reporters/Reporter.h"
#include "Approvals.h"

#include "CombinationApprovals.h" // For Empty

class CombinationApprovalsOld
{
public:
    CombinationApprovalsOld() = delete;
    ~CombinationApprovalsOld() = delete;

    template <
            typename Container1,
            typename Container2,
            typename Container3,
            typename ReturnType>
    static void verifyAllCombinations(
            std::function<ReturnType (
                    typename Container1::value_type,
                    typename Container2::value_type,
                    typename Container3::value_type)> converter,
            const Container1& inputs1,
            const Container2& inputs2,
            const Container3& inputs3,
            const Reporter& reporter = DefaultReporter())
    {
        Empty empty;
        std::stringstream s;
        for (auto input1 : inputs1)
        {
            for (auto input2 : inputs2)
            {
                for (auto input3 : inputs3)
                {
                    ReturnType result = converter(input1, input2, input3);
                    s << "(" << input1;
                    if (empty != input2) { s << ", " << input2; }
                    if (empty != input3) { s << ", " << input3; }
                    s << ") => " << result << '\n';
                }
            }
        }
        Approvals::verify(s.str(), reporter);
    }

    template <
            typename Container1,
            typename Container2,
            typename ReturnType>
    static void verifyAllCombinations(
            std::function<ReturnType (
                    typename Container1::value_type,
                    typename Container2::value_type)> converter,
            const Container1& inputs1,
            const Container2& inputs2,
            const Reporter& reporter = DefaultReporter())
    {
        verifyAllCombinations<
                Container1,
                Container2,
                std::vector<Empty>, ReturnType>([&](
                                                      typename Container1::value_type i1,
                                                      typename Container2::value_type i2,
                                                      Empty){return converter(i1, i2);},
                                              inputs1,
                                              inputs2,
                                              empty(),
                                              reporter);
    }

    template <
            typename Container1,
            typename ReturnType>
    static void verifyAllCombinations(
            std::function<ReturnType (
                    typename Container1::value_type)> converter,
            const Container1& inputs1,
            const Reporter& reporter = DefaultReporter())
    {
        verifyAllCombinations<
                Container1,
                std::vector<Empty>, ReturnType>([&](
                                                      typename Container1::value_type i1,
                                                      Empty){return converter(i1);},
                                              inputs1,
                                              empty(),
                                              reporter);
    }

    // Implementation details: these are left public to allow users
    // to write additional verifyAllCombinations overloads that take more
    // than 9 arguments - unlikely though that may be.
    using EmptyContainer = std::vector<Empty>;
    static EmptyContainer empty()
    {
        return EmptyContainer{Empty()};
    }
};

#endif
