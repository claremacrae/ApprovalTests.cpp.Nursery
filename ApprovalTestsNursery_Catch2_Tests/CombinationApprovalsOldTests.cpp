#include "Catch.hpp"
#include <map>
#include <vector>
#include <string>
#include "PairUtilities.h"
#include "CombinationApprovalsOld.h"
#include "reporters/FakeReporter.h"

TEST_CASE("YouCanVerifyCombinationsOf1") {
    std::vector<std::string> words{"hello", "world"};
    CombinationApprovalsOld::verifyAllCombinations( [](std::string s){return s + "!";}, words);
}

TEST_CASE("YouCanVerifyCombinationsOf1WithTemplateParameters") {
    std::vector<std::string> words{"hello", "world"};
    CombinationApprovalsOld::verifyAllCombinations<std::vector<std::string>, std::string>( [](std::string s){return s + "!";}, words);
}

TEST_CASE("YouCanVerifyCombinationsOf1Reports") {
    std::vector<std::string> words{"hello", "world"};
    FakeReporter reporter;
    try
    {
        CombinationApprovalsOld::verifyAllCombinations( [](std::string s){return s + "!";}, words, reporter);
    }
    catch(const ApprovalException&)
    {
        // ignore
    }
    REQUIRE(reporter.called == true);
}

// begin-snippet: YouCanVerifyCombinationsOf2
TEST_CASE("YouCanVerifyCombinationsOf2") {
    std::vector<std::string> v{"hello", "world"};
    std::vector<int> numbers{1, 2, 3};
    CombinationApprovalsOld::verifyAllCombinations<
        std::vector<std::string>, std::vector<int>, std::pair<std::string, int>>(
            [](std::string s, int i){return std::make_pair(s, i);},
            v,
            numbers);
}
// end-snippet

#if 0
TEST_CASE("YouCanVerifyCombinationsOf9") {
    std::vector<std::string> letters{"a", "b"};
    CombinationApprovalsOld::verifyAllCombinations<
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::string>( [](
            std::string s1,
            std::string s2,
            std::string s3,
            std::string s4,
            std::string s5,
            std::string s6,
            std::string s7,
            std::string s8,
            std::string s9)
                              {return s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8 + s9;}, letters, letters, letters, letters, letters, letters, letters, letters, letters);
}
#endif

TEST_CASE("YouCanVerifyCombinationsOf3") {
    std::vector<std::string> letters{"a", "b"};
    CombinationApprovalsOld::verifyAllCombinations<
            std::vector<std::string>,
            std::vector<std::string>,
            std::vector<std::string>,
            std::string>( [](
            std::string s1,
            std::string s2,
            std::string s3)
                              {return s1 + s2 + s3;}, letters, letters, letters);
}
