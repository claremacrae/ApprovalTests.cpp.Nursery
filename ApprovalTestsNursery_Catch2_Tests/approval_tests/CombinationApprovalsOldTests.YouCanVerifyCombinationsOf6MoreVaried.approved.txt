(a, b, c, d, e, f) => abcdef

